File: cpp/src/y2017/d01.cpp
#include "../Day.hpp"

class Day01 : public Day
{
public:
  Day01(int year, int day) : Day(year, day) {}

  std::string partOne() override
  {
    return std::to_string(calculate(input(), 1));
  }

  std::string partTwo() override
  {
    auto data = input();
    size_t steps = data.length() / 2;
    return std::to_string(calculate(data, steps));
  }

private:
  std::string input() { return readInput(""); }

  static int calculate(std::string input, int steps)
  {
    int sum = 0;
    for (size_t i = 0; i < input.length(); i++)
    {
      size_t next_idx = (i + steps) % input.length();
      if (input[i] == input[next_idx])
        sum += c_to_digit(input[i]);
    }

    return sum;
  }

  static int c_to_digit(char c) { return (c - '0'); }
};

REGISTER_DAY(2017, 1, Day01);

File: python/y2018/d01.py
from day import Day, register_day


@register_day(2018, 1)
class Day01(Day):
    def part_one(self) -> str:
        return str(sum([int(x) for x in self.read_input().split("\n")]))

    def part_two(self) -> str:
        input = [int(x) for x in self.read_input().split("\n")]
        result = None
        frequency = 0
        mem = {}

        while result is None:
            for x in input:
                frequency += x
                if frequency in mem:
                    result = frequency
                    break
                else:
                    mem[frequency] = 1

        return str(result)
